File: CarlaControl/src/carlacontrol/SumoInstallation.py
"""Locate a SUMO installation and the executables and Python tools inside it.

SUMO is not a Python package: its command-line tools, its `traci`/`sumolib` modules and its PROJ data
all live inside an installation directory that SUMO's own convention names with the `SUMO_HOME`
environment variable. This resolves that directory once so callers do not each hard-code a path, and
looks in three places in order:

  1. a path given explicitly by the caller,
  2. `SUMO_HOME`, which is what a SUMO install sets and what `traci` itself checks,
  3. any additional candidate directories the caller offers (a source build inside a repository,
     say), and finally the executable search path.

That order is deliberate and is not changed by the reporting below: matching what raw `traci` resolves
on the same machine is worth more than always preferring a repository's own build, because tooling that
disagrees with `traci` about which SUMO is in use is a worse kind of silent divergence than the one
being reported.

**Every resolution is announced.** More than one SUMO can be installed on a machine, and the
executables give no hint of which one a tool picked. A repository-pinned build converting a map while
an unrelated system-wide install authors the scenarios against it is a real, observed failure: it
produces a network and a scenario that are not from the same converter, and nothing says so. So
`locate` reports the path, the version and *which rule matched* at `INFO`, once per distinct
resolution, with no flag to turn it off, and `require_version` turns a known mismatch against a
recorded converter into a refusal rather than a note in a log nobody reads.
"""
from __future__ import annotations

import logging
import os
import re
import shutil
import subprocess
import sys
from collections.abc import Iterable
from dataclasses import dataclass
from functools import cached_property
from pathlib import Path

EXECUTABLE_SUFFIX = ".exe" if os.name == "nt" else ""

# `Eclipse SUMO netconvert 1.27.0` / `Eclipse SUMO sumo 1.27.1` — the first line of any SUMO tool's
# `--version` output. The trailing dot-separated number is the release; anything after it (a git
# description on a development build, for example) is not part of the comparison.
_VERSION_LINE = re.compile(r"Eclipse SUMO \S+ v?(\d+(?:\.\d+)*)")

# Resolutions already announced this process, keyed by the path and the rule that matched, so a
# long-running tool does not repeat the line while a *second*, different resolution is still reported
# rather than swallowed.
_ANNOUNCED: set[tuple[str, str]] = set()

logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class SumoInstallation:
    """One SUMO installation on this machine, and how it came to be the one in use."""

    home: Path

    #: Which rule in `locate` matched: `explicit`, `SUMO_HOME`, `candidate` or `PATH`. Knowing the
    #: path alone says what resolved but not why, which is the question asked when it is the wrong one.
    source: str = "explicit"

    @classmethod
    def locate(cls, explicit: str | Path | None = None,
               extra_candidates: Iterable[str | Path] = ()) -> SumoInstallation:
        """Find a usable installation, or raise saying where it looked.

        The resolved path, version and matching rule are logged at `INFO` before returning.
        """
        searched: list[Path] = []
        rules: list[tuple[str, str | Path | None]] = [("explicit", explicit),
                                                      ("SUMO_HOME", os.environ.get("SUMO_HOME"))]
        rules += [("candidate", candidate) for candidate in extra_candidates]
        for source, candidate in rules:
            if not candidate:
                continue
            home = Path(candidate)
            searched.append(home)
            if cls._is_installation(home):
                return cls(home, source)._announce()

        # Nothing declared one, so fall back to a `sumo` on the executable search path and take
        # the directory above it as the installation.
        found = shutil.which("sumo") or shutil.which("netconvert")
        if found:
            home = Path(found).resolve().parent.parent
            if cls._is_installation(home):
                return cls(home, "PATH")._announce()

        raise FileNotFoundError(
            "no SUMO installation found. Set SUMO_HOME to the directory holding SUMO's bin/ and "
            "tools/ (for example C:\\Program Files (x86)\\Eclipse\\Sumo), put SUMO's bin on PATH, "
            "or pass one explicitly. Looked in: "
            + (", ".join(str(p) for p in searched) or "nothing was given"))

    @staticmethod
    def _is_installation(home: Path) -> bool:
        """A directory counts when it holds at least one SUMO executable."""
        return any((home / "bin" / f"{name}{EXECUTABLE_SUFFIX}").exists()
                   for name in ("sumo", "netconvert"))

    def _announce(self) -> SumoInstallation:
        """Report this resolution once, then return self so `locate` can return it inline."""
        key = (str(self.home), self.source)
        if key in _ANNOUNCED:
            return self
        _ANNOUNCED.add(key)
        message = (f"SUMO installation: {self.home} (version {self.version or 'unknown'}, "
                   f"matched by {self.source})")
        logger.info(message)
        # This line exists so that nobody has to already suspect a mismatch to find out about one, so
        # it must survive a caller that never configured logging at all.
        if not logger.isEnabledFor(logging.INFO):
            print(message, file=sys.stderr)
        return self

    @cached_property
    def version(self) -> str | None:
        """The SUMO release of this installation, e.g. `1.27.0`, or None when it cannot be read.

        Read from `netconvert`, not `sumo`: `_is_installation` deliberately accepts a directory
        holding only `netconvert` (the converter is all some callers need), so asking `sumo` would
        report nothing for an installation that is perfectly usable.
        """
        netconvert = self.home / "bin" / f"netconvert{EXECUTABLE_SUFFIX}"
        if not netconvert.exists():
            return None
        try:
            completed = subprocess.run([str(netconvert), "--version"], capture_output=True,
                                       text=True, timeout=30, check=False)
        except OSError as error:
            logger.warning("could not run %s to read its version: %s", netconvert, error)
            return None
        match = _VERSION_LINE.search(completed.stdout or completed.stderr or "")
        if not match:
            logger.warning("could not parse a version from %s --version", netconvert)
            return None
        return match.group(1)

    def require_version(self, expected: str | None, allow_mismatch: bool = False) -> None:
        """Refuse when this installation is not the SUMO release `expected`.

        `expected` is the version recorded by whatever produced the artifact being worked on — a
        world package records the netconvert that built it. Pass None when nothing recorded one: that
        warns and proceeds, because an older artifact that predates the recording is not evidence of a
        mismatch. `allow_mismatch` warns and proceeds for a caller that has a reason to accept the
        risk, and is what a `--allow-version-mismatch` flag sets.
        """
        if expected is None:
            logger.warning(
                "nothing records which SUMO produced the artifact being worked on, so the "
                "installation at %s (version %s) cannot be checked against it. Proceeding; rebuild "
                "the world to record its converter.", self.home, self.version or "unknown")
            return

        resolved = self.version
        if resolved == expected:
            return

        detail = (f"this world was built with SUMO {expected}; the installation at {self.home} "
                  f"(matched by {self.source}) is "
                  + (f"SUMO {resolved}" if resolved else "of an unreadable version"))
        if allow_mismatch:
            logger.warning("%s. Proceeding because the version mismatch was explicitly allowed.",
                           detail)
            return
        raise RuntimeError(
            f"{detail}. Two SUMO releases are not guaranteed to convert the same OSM into the same "
            "network. Pass --sumo-home to point at a matching installation, rebuild the world with "
            "this one, or pass --allow-version-mismatch to accept the risk.")

    def executable(self, name: str) -> Path:
        path = self.home / "bin" / f"{name}{EXECUTABLE_SUFFIX}"
        if not path.exists():
            raise FileNotFoundError(f"{name} is missing from {self.home / 'bin'}")
        return path

    @property
    def sumo(self) -> Path:
        return self.executable("sumo")

    @property
    def sumo_gui(self) -> Path:
        return self.executable("sumo-gui")

    @property
    def netconvert(self) -> Path:
        return self.executable("netconvert")

    @property
    def duarouter(self) -> Path:
        """The route validator, which scenario compilation runs on every authored route."""
        return self.executable("duarouter")

    @property
    def tools(self) -> Path:
        """The directory holding `traci` and `sumolib`."""
        return self.home / "tools"

    @property
    def proj_data(self) -> Path | None:
        """PROJ's data directory, which netconvert needs to reproject. None when absent."""
        share = self.home / "share" / "proj"
        return share if (share / "proj.db").exists() else None

    def add_tools_to_path(self) -> None:
        """Make `traci` and `sumolib` importable from this installation."""
        tools = str(self.tools)
        if tools not in sys.path:
            sys.path.insert(0, tools)

    def import_traci(self):
        """Import `traci` from this installation, reporting it when it comes from somewhere else."""
        self.add_tools_to_path()
        try:
            import traci
        except ImportError as error:
            raise ImportError(
                f"traci was not found under {self.tools}. The installation at {self.home} looks "
                "incomplete; point SUMO_HOME at one that includes its tools directory.") from error
        # An installation with no tools/ directory, or a traci already imported by something else,
        # leaves Python free to satisfy the import from another SUMO on PYTHONPATH -- so the module
        # driving the simulation can come from a different release than the binaries running it.
        origin = getattr(traci, "__file__", None)
        if origin and self.tools.resolve() not in Path(origin).resolve().parents:
            logger.warning("traci was imported from %s, which is not under %s; the Python modules "
                           "and the SUMO binaries in use are from different installations.",
                           origin, self.tools)
        return traci

File: CarlaNet/src/CarlaNet.Sumo/SumoInstallation.cs
using System.Diagnostics;
using System.Text.RegularExpressions;

namespace CarlaNet.Sumo;

/// <summary>
/// One SUMO installation on this machine, and which rule found it.
/// </summary>
/// <remarks>
/// <para>SUMO is not a NuGet package. Its executables live in an installation directory that SUMO's
/// own convention names with the <c>SUMO_HOME</c> environment variable, and something has to find
/// that directory before <c>sumo</c> can be launched.</para>
///
/// <para><b>The repository's own pinned build is preferred over <c>SUMO_HOME</c>, and that is
/// deliberate.</b> The identifiers in <see cref="TraCIConstants"/> were translated from one SUMO
/// release, and this machine already has two SUMO installations of different releases on it. A
/// protocol client survives that far better than a linked binding would -- a mismatch is reported
/// by the handshake rather than being undefined -- but "reported" still means a run that had to be
/// thrown away, so the default resolves to the release the client was written for. A distribution
/// recipient has no repository build at all, which is where <c>SUMO_HOME</c> takes over, and it is
/// how the launcher scripts already point at the bundled toolchain.</para>
///
/// <para>Every resolution records <see cref="Source"/>, because more than one SUMO can be installed
/// and the binaries give no hint of which one a process picked.</para>
/// </remarks>
public sealed class SumoInstallation
{
    /// <summary>
    /// Environment variable naming an installation root outright, ahead of every other rule. The
    /// escape hatch for a machine where neither the repository build nor <c>SUMO_HOME</c> is the
    /// one wanted.
    /// </summary>
    public const string OverrideVariable = "CARLANET_SUMO_HOME";

    /// <summary>SUMO's own convention for the root holding <c>bin/</c> and <c>tools/</c>.</summary>
    public const string HomeVariable = "SUMO_HOME";

    /// <summary>Where <c>CarlaSetup</c> stages the pinned toolchain, relative to the repository root.</summary>
    private const string StagedInstallation = "Build/sumo-install";

    /// <summary>
    /// Where SUMO's own build writes its binaries before <c>CarlaSetup</c> stages them. Checked
    /// after the staged copy, so a build interrupted before staging still resolves to the pinned
    /// release rather than falling through to an unrelated system-wide one.
    /// </summary>
    private const string SourceBuild = "Build/sumo-src";

    // `Eclipse SUMO sumo 1.27.0` -- the first line of any SUMO tool's --version output. A
    // development build appends a git description, which is not part of the release number.
    private static readonly Regex VersionLine = new(@"Eclipse SUMO \S+ v?(\d+(?:\.\d+)*)",
                                                    RegexOptions.Compiled);

    private static readonly object LocateGate = new();
    private static SumoInstallation? _located;
    private static bool _locateAttempted;
    private static string[] _searched = [];

    private string? _release;
    private bool _releaseProbed;

    private SumoInstallation(string home, string source)
    {
        Home = home;
        Source = source;
    }

    /// <summary>The installation root, holding <c>bin/</c> and, in a complete installation, <c>tools/</c>.</summary>
    public string Home { get; }

    /// <summary>
    /// Which rule matched: <c>CARLANET_SUMO_HOME</c>, <c>staged</c>, <c>source-build</c>,
    /// <c>SUMO_HOME</c> or <c>PATH</c>. The path alone says what resolved but not why, which is the
    /// question asked when it is the wrong one.
    /// </summary>
    public string Source { get; }

    /// <summary>The directory holding <c>sumo</c>, <c>duarouter</c> and <c>netconvert</c>.</summary>
    public string BinaryDirectory => Path.Combine(Home, "bin");

    /// <summary>
    /// The directory holding SUMO's Python tools, including its own reference TraCI client. Present
    /// in a complete installation and absent from a bare binary directory, which is why the
    /// constants check that reads it skips rather than fails when it is missing.
    /// </summary>
    public string ToolsDirectory => Path.Combine(Home, "tools");

    /// <summary>The microsimulation binary, launched as a child process and spoken to over TraCI.</summary>
    public string Sumo => Executable("sumo");

    /// <summary>The route validator scenario authoring runs.</summary>
    public string Duarouter => Executable("duarouter");

    /// <summary>The network converter the world pipeline runs.</summary>
    public string Netconvert => Executable("netconvert");

    /// <summary>
    /// The SUMO release this installation reports, or <see langword="null"/> when <c>sumo</c> could
    /// not be run. Probed once, from <c>sumo --version</c>.
    /// </summary>
    public string? Release
    {
        get
        {
            if (_releaseProbed)
            {
                return _release;
            }

            _releaseProbed = true;
            _release = ReadRelease(Sumo);
            return _release;
        }
    }

    /// <summary>Full path to one of SUMO's executables in this installation.</summary>
    public string Executable(string name) =>
        Path.Combine(BinaryDirectory, OperatingSystem.IsWindows() ? name + ".exe" : name);

    /// <summary>
    /// The installation this process will use, or <see langword="null"/> when none resolves.
    /// Resolved once and cached; <see cref="SearchedDirectories"/> then says where it looked.
    /// </summary>
    public static SumoInstallation? Locate()
    {
        lock (LocateGate)
        {
            if (_locateAttempted)
            {
                return _located;
            }

            _locateAttempted = true;
            List<string> searched = [];
            _located = Search(searched);
            _searched = [.. searched];
            return _located;
        }
    }

    /// <summary>
    /// The installation this process will use, or a <see cref="DirectoryNotFoundException"/> naming
    /// every directory that was tried.
    /// </summary>
    public static SumoInstallation LocateOrThrow() =>
        Locate() ?? throw new DirectoryNotFoundException(
            "No SUMO installation was found. Run CarlaSetup to stage one under "
            + $"{StagedInstallation}, set {HomeVariable} or {OverrideVariable} to an installation's "
            + "root, or put SUMO's bin directory on the executable search path. Looked in: "
            + (SearchedDirectories.Count == 0 ? "nothing was given" : string.Join(", ", SearchedDirectories)));

    /// <summary>Every directory <see cref="Locate"/> tried, in order. Empty until it has run once.</summary>
    public static IReadOnlyList<string> SearchedDirectories => _searched;

    private static SumoInstallation? Search(List<string> searched)
    {
        foreach ((string source, string? home) in Candidates())
        {
            if (string.IsNullOrWhiteSpace(home))
            {
                continue;
            }

            string full = Path.GetFullPath(home);
            searched.Add(full);
            string executable = Path.Combine(full, "bin",
                                             OperatingSystem.IsWindows() ? "sumo.exe" : "sumo");
            if (File.Exists(executable))
            {
                return new SumoInstallation(full, source);
            }
        }

        return null;
    }

    private static IEnumerable<(string Source, string? Home)> Candidates()
    {
        yield return (OverrideVariable, Environment.GetEnvironmentVariable(OverrideVariable));

        // Both repository locations hold the release the constant table was translated from.
        yield return ("staged", FindUpwards(StagedInstallation));
        yield return ("source-build", FindUpwards(SourceBuild));

        yield return (HomeVariable, Environment.GetEnvironmentVariable(HomeVariable));

        yield return ("PATH", FindOnPath());
    }

    /// <summary>
    /// Walk from the running assembly towards the filesystem root looking for a directory that
    /// holds <paramref name="relative"/>. The build output sits several levels below the repository
    /// root and deeper still inside a git worktree, so the depth cannot be hard-coded.
    /// </summary>
    private static string? FindUpwards(string relative)
    {
        foreach (string start in SearchRoots())
        {
            DirectoryInfo? directory = new(start);
            while (directory is not null)
            {
                string candidate = Path.Combine(directory.FullName, relative);
                if (Directory.Exists(candidate))
                {
                    return candidate;
                }

                directory = directory.Parent;
            }
        }

        return null;
    }

    /// <summary>
    /// Where the upward walk starts: the application's base directory, and this assembly's own.
    /// </summary>
    /// <remarks>
    /// Two starting points, and the second is not redundant. An application's base directory is the
    /// executable's, and <b>there is no executable when the runtime is hosted</b> -- loaded into
    /// another process, as it is when a Python orchestrator drives the bridge,
    /// <see cref="AppContext.BaseDirectory"/> is the empty string and constructing a
    /// <see cref="DirectoryInfo"/> from it raises. The assembly's own directory is under the
    /// repository in exactly that case, which is the case where the repository build is the
    /// installation wanted.
    /// </remarks>
    internal static IEnumerable<string> SearchRoots()
    {
        string applicationBase = AppContext.BaseDirectory;
        if (!string.IsNullOrWhiteSpace(applicationBase))
        {
            yield return applicationBase;
        }

        string assembly = Path.GetDirectoryName(typeof(SumoInstallation).Assembly.Location)
                          ?? string.Empty;
        if (!string.IsNullOrWhiteSpace(assembly)
            && !string.Equals(Path.TrimEndingDirectorySeparator(assembly),
                              Path.TrimEndingDirectorySeparator(applicationBase),
                              StringComparison.OrdinalIgnoreCase))
        {
            yield return assembly;
        }
    }

    /// <summary>
    /// A <c>sumo</c> on the executable search path, reported as the installation root above the
    /// directory holding it.
    /// </summary>
    private static string? FindOnPath()
    {
        string? path = Environment.GetEnvironmentVariable("PATH");
        if (path is null)
        {
            return null;
        }

        string executable = OperatingSystem.IsWindows() ? "sumo.exe" : "sumo";
        foreach (string entry in path.Split(Path.PathSeparator))
        {
            if (entry.Length > 0 && File.Exists(Path.Combine(entry, executable)))
            {
                return Path.GetDirectoryName(Path.GetFullPath(entry));
            }
        }

        return null;
    }

    private static string? ReadRelease(string executable)
    {
        if (!File.Exists(executable))
        {
            return null;
        }

        try
        {
            using Process? process = Process.Start(new ProcessStartInfo(executable, "--version")
            {
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true,
            });

            if (process is null)
            {
                return null;
            }

            string output = process.StandardOutput.ReadToEnd();
            process.WaitForExit();
            Match match = VersionLine.Match(output);
            return match.Success ? match.Groups[1].Value : null;
        }
        catch (Exception exception) when (exception is IOException
                                              or System.ComponentModel.Win32Exception
                                              or InvalidOperationException)
        {
            return null;
        }
    }
}
